### DISCUSSION TITLE
Error Encountered: Cannot Read Properties of Undefined (Reading 'map') When Running the Deploy and FixedKey Commands

### DISCUSSION THREAD

**Discussion by haolt-toratech** [Accenture/sfmc-devtools#1933] [Q&A]
Hello MC Dev Team,

I’m new to both your tool and Salesforce Marketing Cloud, and I would greatly appreciate your help with the issue I’m encountering.

#### Goal
My goal is to deploy changes to an SQL query file and the content of an automation verification step to Marketing Cloud by running the `deploy` command, as well as the `fixedKeys` command for queries and DataExtensions as the files are named using external keys. However, I have been unable to successfully execute these commands. Below is an image of the error I encountered:

![2](https://github.com/user-attachments/assets/a0a9dfc3-b18c-4867-904b-c193beb3d24c)
Note: A similar error was displayed during different command processes.

#### Issue

The problem seems to be related to DataExtensions and specifically subtypes such as:
- shared_data
- shared_dataextension
- shared_salesforcedataextension
- synchronizeddataextension

Whenever I run commands that involve processing or caching these subtypes (e.g., mcdev deploy or mcdev fixedKeys), the process fails.  This issue occurs for resources such as:

- Automation verification steps
- SQL query files
- DataExtensions themselves

Could you please advise on how to resolve this issue? Specifically:
1. How can I skip or bypass the caching process for these subtypes during commands like deploy or fixedKeys?
3. Are there additional configurations or adjustments I need to make in my config.json or elsewhere to avoid this error?

#### Details
1. Initialy, I was able to retrieve all resources, except `dataExtension` and `query`,  from my Marketing Cloud instance to local folder using the command:

> mcdev retrieve [BU]

I've noticed encountered this error in the log:
>error: - Retrieving dataExtension failed: Cannot read properties of undefined (reading 'map')

Interestingly, if I run the command specifically for DataExtensions or query, it retrieves those resource successfully without errors, even the caching step is went through, no error log was displayed:
>mcdev retrieve [BU] dataExtension  

2. Running the deploy or fixedKey command failed during the caching subtypes step:

deploy command: 
>mcdev deploy [BU] query "key_name" --fromRetrieve --debug

fixKey command: 
>mcdev fixKeys [BU] query

Same error was displayed: error: mcdev.deploy failed: Cannot read properties of undefined (reading 'map')

> 20:10:08 error: mcdev.deploy failed: Cannot read properties of undefined (reading 'map')
> 20:10:08 debug: TypeError: Cannot read properties of undefined (reading 'map')
>     at Folder.retrieve (file:///opt/homebrew/lib/node_modules/mcdev/lib/metadataTypes/Folder.js:58:52)
>     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
>     at async DataExtension.retrieveSharedForCache (file:///opt/homebrew/lib/node_modules/mcdev/lib/metadataTypes/DataExtension.js:830:24)
>     at async DataExtension.retrieve (file:///opt/homebrew/lib/node_modules/mcdev/lib/metadataTypes/DataExtension.js:776:38)
>     at async Deployer._deploy (file:///opt/homebrew/lib/node_modules/mcdev/lib/Deployer.js:347:28)
>     at async Deployer._deployBU (file:///opt/homebrew/lib/node_modules/mcdev/lib/Deployer.js:248:40)
>     at async Deployer.deploy (file:///opt/homebrew/lib/node_modules/mcdev/lib/Deployer.js:204:46)
>     at async Mcdev.deploy (file:///opt/homebrew/lib/node_modules/mcdev/lib/index.js:507:30)

**Comment by JoernBerkefeld** [comment]
lets track that under the bug ticket because even if this is instance / BU specific and doesn't happen for others, the solution should be made more reliable to handle it gracefully.
- #1932
